File: apps/mobile/lib/sse.dart
// sse.dart — streaming del chat (POST /v1/chat/completions con stream:true).
// Mismo protocolo que la web (lib/stream.js): chunks formato OpenAI,
// comentarios keep-alive y cierre con "data: [DONE]".
import 'dart:convert';

import 'package:http/http.dart' as http;

import 'api.dart';

class ChatStreamHandle {
  final http.Client _client;
  bool cancelled = false;

  ChatStreamHandle(this._client);

  void cancel() {
    cancelled = true;
    _client.close();
  }
}

ChatStreamHandle streamChatCompletion({
  required String base,
  required String? token,
  required String model,
  required List<Map<String, dynamic>> messages,
  required String conversationId,
  required bool webSearch,
  required void Function(String delta) onDelta,
  required void Function(List<dynamic> sources) onSources,
  required void Function() onDone,
  required void Function(Object error) onError,
}) {
  final client = http.Client();
  final handle = ChatStreamHandle(client);

  Future<void> run() async {
    try {
      final req = http.Request('POST', Uri.parse('$base/v1/chat/completions'));
      req.headers['Content-Type'] = 'application/json';
      req.headers['Accept'] = 'text/event-stream';
      if (token != null) req.headers['Authorization'] = 'Bearer $token';
      req.body = jsonEncode({
        'model': model,
        'messages': messages,
        'conversation_id': conversationId,
        'stream': true,
        'web_search': webSearch,
        // Historial compartido con la web: misma cuenta, mismas conversaciones.
        'source': 'web',
      });

      final res = await client.send(req);
      if (res.statusCode >= 400) {
        final body = await res.stream.bytesToString();
        dynamic data;
        try {
          data = jsonDecode(body);
        } catch (_) {
          data = null;
        }
        throw ApiException(extractDetail(data, res.statusCode), status: res.statusCode);
      }

      var buffer = '';
      var done = false;
      await for (final chunk in res.stream.transform(utf8.decoder)) {
        if (handle.cancelled) return;
        buffer += chunk;
        final events = buffer.split('\n\n');
        buffer = events.removeLast(); // el último puede venir incompleto
        for (final event in events) {
          final line = event.trim();
          if (!line.startsWith('data:')) continue; // ignora keep-alives
          final data = line.substring(5).trim();
          if (data == '[DONE]') {
            done = true;
            break;
          }
          try {
            final obj = jsonDecode(data);
            if (obj is Map && obj['lixbon_sources'] is List) {
              onSources(obj['lixbon_sources'] as List);
              continue;
            }
            if (obj is Map) {
              final choices = obj['choices'];
              if (choices is List && choices.isNotEmpty) {
                final delta = choices[0]?['delta']?['content'];
                if (delta is String && delta.isNotEmpty) onDelta(delta);
              }
            }
          } catch (_) {
            // chunk malformado: se ignora
          }
        }
        if (done) break;
      }
      if (!handle.cancelled) onDone();
    } catch (err) {
      if (!handle.cancelled) onError(err);
    } finally {
      client.close();
    }
  }

  run();
  return handle;
}

File: apps/mobile/src/sse.js
// sse.js — streaming del chat (POST /v1/chat/completions con stream:true).
// Mismo protocolo que la web (lib/stream.js): chunks formato OpenAI,
// comentarios keep-alive y cierre con "data: [DONE]". El fetch de React
// Native no expone el cuerpo incremental, así que se usa XMLHttpRequest
// leyendo responseText en onprogress.
import { ApiException, extractDetail } from './api';

export function streamChatCompletion({
  base,
  token,
  model,
  messages,
  conversationId,
  webSearch,
  onDelta,
  onSources,
  onDone,
  onError,
}) {
  const xhr = new XMLHttpRequest();
  let cancelled = false;
  let finished = false;
  let seen = 0; // responseText ya procesado (el XHR lo acumula entero)
  let buffer = '';

  const finishOk = () => {
    if (!cancelled && !finished) {
      finished = true;
      onDone();
    }
  };
  const fail = (err) => {
    if (!cancelled && !finished) {
      finished = true;
      onError(err);
    }
  };

  const pump = (isFinal) => {
    const text = xhr.responseText || '';
    if (text.length > seen) {
      buffer += text.slice(seen);
      seen = text.length;
    }
    let idx;
    while ((idx = buffer.indexOf('\n\n')) !== -1) {
      const event = buffer.slice(0, idx).trim();
      buffer = buffer.slice(idx + 2);
      if (!event.startsWith('data:')) continue; // ignora keep-alives
      const data = event.slice(5).trim();
      if (data === '[DONE]') {
        finishOk();
        return;
      }
      try {
        const obj = JSON.parse(data);
        if (obj && Array.isArray(obj.lixbon_sources)) {
          onSources(obj.lixbon_sources);
          continue;
        }
        const delta = obj?.choices?.[0]?.delta?.content;
        if (typeof delta === 'string' && delta) onDelta(delta);
      } catch {
        // chunk malformado: se ignora
      }
    }
    if (isFinal) finishOk();
  };

  xhr.open('POST', `${base}/v1/chat/completions`);
  xhr.setRequestHeader('Content-Type', 'application/json');
  xhr.setRequestHeader('Accept', 'text/event-stream');
  if (token) xhr.setRequestHeader('Authorization', `Bearer ${token}`);
  xhr.timeout = 0; // el stream puede estar minutos abierto

  xhr.onprogress = () => {
    if (!cancelled && xhr.status < 400) pump(false);
  };
  xhr.onload = () => {
    if (cancelled || finished) return;
    if (xhr.status >= 400) {
      let data = null;
      try {
        data = JSON.parse(xhr.responseText);
      } catch {
        // cuerpo no JSON
      }
      fail(new ApiException(extractDetail(data, xhr.status), xhr.status));
      return;
    }
    pump(true);
  };
  xhr.onerror = () => fail(new ApiException('Se perdió la conexión con el servidor'));

  xhr.send(
    JSON.stringify({
      model,
      messages,
      conversation_id: conversationId,
      stream: true,
      web_search: webSearch,
      // Historial compartido con la web: misma cuenta, mismas conversaciones.
      source: 'web',
    }),
  );

  return {
    cancel() {
      cancelled = true;
      try {
        xhr.abort();
      } catch {
        // ya cerrado
      }
    },
  };
}
